overlord/snapstate/aliases.go

```diff
@@ -0,0 +1,338 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package snapstate
+
+import (
+	"encoding/json"
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"gopkg.in/tomb.v2"
+
+	"github.com/snapcore/snapd/i18n/dumb"
+	"github.com/snapcore/snapd/overlord/snapstate/backend"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/snap"
+)
+
+type AliasState struct {
+	// Enabled tracks for which snap the alias has been enabled, "" if none
+	Enabled string `json:"enabled,omitempty"`
+	// Auto is true if the alias has been auto-enabled
+	// TODO: Auto bool `json:"auto,omitempy"`
+	// Disabled tracks the possibly many snap for which the alias has
+	// been forcefully disabled
+	// TODO: Disabled []string `json:"disabled,omitempty"`
+}
+
+type aliasStateMap map[string]*AliasState
+
+func getAliases(st *state.State) (aliasStateMap, error) {
+	var aliasStates map[string]*AliasState
+	err := st.Get("aliases", &aliasStates)
+	if err != nil && err != state.ErrNoState {
+		return nil, err
+	}
+	return aliasStates, nil
+}
+
+func (m aliasStateMap) get(alias string) AliasState {
+	var aliasState AliasState
+	if got := m[alias]; got != nil {
+		aliasState = *got
+	}
+	return aliasState
+}
+
+func (m aliasStateMap) set(alias string, aliasState *AliasState) {
+	if *aliasState != (AliasState{}) {
+		m[alias] = aliasState
+	} else {
+		delete(m, alias)
+	}
+}
+
+// Alias enables the aliases for snap
+func Alias(st *state.State, snapName string, aliases []string) (*state.TaskSet, error) {
+	var snapst SnapState
+	err := Get(st, snapName, &snapst)
+	if err == state.ErrNoState {
+		return nil, fmt.Errorf("cannot find snap %q", snapName)
```

Comment:
Not sure if that's true or not for native speakers, but "%q snap" feels slightly nicer than "snap %q" for some reason.